core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt

```diff
@@ -461,48 +461,73 @@ object GeneUtils {
     /**
      * force at least one boolean to be selected
      */
-    fun repairBooleanSelection(obj: ObjectGene) {
+    fun repairBooleanSelection(obj: ObjectGene): Boolean {
 
         if (obj.fields.isEmpty()
             || obj.fields.count { it !is OptionalGene && it !is BooleanGene && it !is TupleGene } > 0
         ) {
             throw IllegalArgumentException("There should be at least 1 field, and they must be all optional or boolean or tuple")
         }
 
+        /*
+        * Filtering functional genes among all genes.
+        * */
         val selected = obj.fields.filter {
             ((it is OptionalGene && it.isActive) ||
                     (it is BooleanGene && it.value) ||
-                    (it is OptionalGene && it.gene is TupleGene && it.gene.lastElementTreatedSpecially && isLastSelected(it.gene)) ||
+                    (it is OptionalGene && it.gene is TupleGene && it.gene.lastElementTreatedSpecially && isLastSelected(
+                        it.gene
+                    )) ||
                     (it is TupleGene && it.lastElementTreatedSpecially && isLastSelected(it))
                     )
         }
 
+        /*
+        * we have found some functional genes.
+        * we need to look into them to find functional genes as well
+        * */
         if (selected.isNotEmpty()) {
             //it is fine, but we still need to make sure selected objects are fine
             selected.forEach {
                 if ((it is OptionalGene && it.gene is ObjectGene && (it.gene !is CycleObjectGene || it.gene !is LimitObjectGene))
```

Comment:
`CycleObjectGene` and `LimitObjectGene` no longer extends `ObjectGene`, so these checks seem redundant

Reply:
Removed